### PR TITLE
Offering default methods on ParameterizedJob

### PR DESCRIPTION
**Pull request by jglick**
# Description

Allows subtypes to have less boilerplate. Mentioned as a design pattern in #2730.

### Changelog entries

None required.

### Desired reviewers

@reviewbybees

### CODE REVIEW

File: core/src/main/java/jenkins/model/ParameterizedJobMixIn.java
Comment (jglick) [inline core/src/main/java/jenkins/model/ParameterizedJobMixIn.java]: cf. https://github.com/stapler/stapler/pull/113

**Review by abayer — APPROVED** [review]

**Comment by jglick** [comment]
```
[ERROR] …/core/src/main/java/jenkins/model/ParameterizedJobMixIn.java:165: error: reference not found
[ERROR] * Standard implementations of {@link ParameterizedJob#isParametrized}.
[ERROR] ^
[ERROR] 
```

File: core/src/main/java/jenkins/model/ParameterizedJobMixIn.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/ParameterizedJobMixIn.java]: So, no Restricted/RequirePOST annotations here? I am not sure it is safe to offer such default impl. 🐛  , but probably I miss something.

File: core/src/main/java/jenkins/model/ParameterizedJobMixIn.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/ParameterizedJobMixIn.java]: same

**Review by oleg-nenashev — DISMISSED** [review]
Kinda 🐛 since it advertises non-Restricted & non-RequirePOST method in all `ParameterizedJob` implementations. Even though it is a historic API in AbstractProject, I am not sure we want to offer it in other job types

**Review by stephenc — APPROVED** [review]
Looks good to me.

The public `doBuild` method that is being moved does not currently have any annotations, so while I think it should have a `@RequirePOST` annotation, this refactoring is *not making the code worse than it already is* and therefore the lack of an annotation is not blocking this PR IMHO

**Comment by oleg-nenashev** [comment]
@stephenc 

> ... this refactoring is not making the code worse than it already is and therefore the lack of an annotation is not blocking this PR IMHO

From what I see it adds new API to the ParameterizedJob interface (new default method), so IMHO it makes the code worse.

**Comment by stephenc** [comment]
> From what I see it adds new API to the ParameterizedJob interface (new default method doBuilds(), which was not defined in the interface before), so IMHO it makes the code worse.

Well `ParameterizedJob` is a marker interface and this is just a refactoring of methods in classes implementing that marker interface, so I see it as non-blocking *for this PR* but the lack of annotation should be fixed (just in a different PR with a separate JIRA to track that change as I think the annotation change could affect users that may have been accidentally using the "feature" so to co-mingle with this PR would hide the change)

File: core/src/main/java/jenkins/model/ParameterizedJobMixIn.java
Comment (jglick) [inline core/src/main/java/jenkins/model/ParameterizedJobMixIn.java]: Actually as per [this tip](http://stackoverflow.com/a/19976615/12916) it _could_ be defined here. Implementing classes would still need to define the method, but at least they could select a known `super` implementation, which is probably more discoverable than looking for similar methods on a mixin.

File: core/src/main/java/jenkins/model/ParameterizedJobMixIn.java
Comment (jglick) [inline core/src/main/java/jenkins/model/ParameterizedJobMixIn.java]: This case would still be tricky however, since the defining method is `protected` and we do not really want to force implementations to override it as `public`.

File: core/src/main/java/jenkins/model/ParameterizedJobMixIn.java
Comment (jglick) [inline core/src/main/java/jenkins/model/ParameterizedJobMixIn.java]: The original had neither annotation. `@Restricted` cannot be added compatibly. `@RequirePOST` is unwanted since the implementation checks the request method explicitly.

**Comment by jglick** [comment]
Test failures are bogus.

Withdrawing for the moment since I want to check if I can offer `default` versions of even (`public`) methods defined in supertypes.

**Comment by jglick** [comment]
Hmm,

    Failed to execute goal com.infradna.tool:bridge-method-injector:1.15:process (default) on project jenkins-core: Failed to process @WithBridgeMethods: Failed to process …/core/target/classes/hudson/model/AbstractProject.class: INVOKESPECIAL/STATIC on interfaces require ASM 5

seems like exactly what https://github.com/infradna/bridge-method-injector/pull/14 should have been solving. Investigating.

**Comment by jglick** [comment]
BTW test failures in this & related PRs are due to [INFRA-1176](https://issues.jenkins-ci.org/browse/INFRA-1176); will make sure tests pass before merging.

**Review by abayer — APPROVED** [review]

**Review by oleg-nenashev — COMMENTED** [review]
I have checked all usages of "instanceof BuildableItem". Actually there are extra implementations in tests, but I do not see anything risky within jenkinsci org and several other repos I usually check. And yes, the ParameterizedJobMixIn seems to be always buildable according to Javadoc

Here is a list of ones which attracted my attention:

* SAFE - https://github.com/jenkinsci/github-plugin/blob/93d40692ff3866705175624e93ec584d4ac88132/src/main/java/org/jenkinsci/plugins/github/util/JobInfoHelpers.java#L47-L57  => 'item instanceof Job ' check leads to another branch
* LIKELY SAFE - https://github.com/jenkinsci/blueocean-plugin/blob/7416f58310e3a72a24d93a95e435d1db9cbc9d2d/blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/rest/QueueUtil.java#L59
  * invokes doomsday error otherwise. likely the BO code should just invoke `Job#isBuildable()`
  * CC @i386 just in case
* SAFE - https://github.com/i-m-c/jenkins-inheritance-plugin (my favorite source of edge cases outside jenkinsci) - implements Project, which is already buildable
* LIKELY SAFE: Trigger plugins. From what I see, triggers will become `isApplicable()` to BuildableItem instances
* SAFE - https://github.com/jenkinsci/computer-queue-plugin/blob/4c2091b021960c7ac1d73a7bd504ef1ee3208831/src/main/java/org/jenkinsci/plugins/computerqueue/ComputerQueue.java#L24 , no idea what the item could be doing there if it is not buildable

**Review by oleg-nenashev — APPROVED** [review]
🐝 after the review of possible regressions due to making all `ParameterizedJob`s `BuildableItem`s.

@reviewbybees done, passing to @jenkinsci/code-reviewers

File: core/src/main/java/jenkins/model/ParameterizedJobMixIn.java
Comment (jglick) [inline core/src/main/java/jenkins/model/ParameterizedJobMixIn.java]: Well the mixin Javadoc does say

> be scheduled in various ways

which I guess is enough. Anyway `BuildableItem` is an interface which is not used all that much and not especially interesting to people grokking this code.

File: core/src/main/java/hudson/model/AbstractProject.java
Comment (jglick) [inline core/src/main/java/hudson/model/AbstractProject.java]: `implements BuildableItem` is now redundant, could be removed I guess.

> I have checked all usages of "instanceof BuildableItem".

The only implementations of `ParameterizedJob` were already `BuildableItem`s so retrofitting it to implement that interface should not cause any behavioral change even if someone _were_ doing that `instanceof` check.

**Comment by oleg-nenashev** [comment]
According to the discussion at the governance meeting on May 10, merging it towards 2.61

**Pull request merged**
